Apache-Beam-Jobs.md

```diff
@@ -227,6 +229,257 @@ error_pcoll = (
+        | 'Get all ExplorationModels' >> ndb_io.GetModels(
+            exp_models.ExplorationModel.get_all(deleted=False))
+    )
+
+    state_count_pcoll = (
+        exp_model_pcoll
+        | 'Count states' >> beam.Map(self.get_number_of_states)
+    )
+
+    state_count_sum_pcoll = (
+        state_count_pcoll
+        | 'Sum values' >> beam.CombineGlobally(sum)
+    )
+```
+
+`CombineGlobally` uses the input function to combine the values in a `PCollection`. It returns a single-element `PCollection`, one which holds the result of the combination.
+
+With this, our objective is complete. However, there's still more code to write!
+
+### 3. Have the `run()` method return a `PCollection` of `JobRunResult`s
+
+* In English, this means that **the job _must_ report _something_ about what occurred during its execution.** For example, this can be the errors it discovered or the number of successful operations it was able to perform. **Empty results are forbidden!**
+
+  * If you don't think your job has any results worth reporting, then just print a "success" metric with the number of models it processed.
+
+* `JobRunResult` outputs should answer the following questions:
+
+  * Did the job run without any problems? How and why do I know?
+  * How much work did the job manage to do?
+  * If the job encountered a problem, what caused it?
+
+Our job is trying to report the total number of states across all explorations, so we need to create `JobRunResult` that holds that information. For this, we can simply use the `as_stdout` helper method:
+
+```python
+def run(self):
+    exp_model_pcoll = (
+        self.pipeline
+        | 'Get all ExplorationModels' >> ndb_io.GetModels(
+            exp_models.ExplorationModel.get_all(deleted=False))
+    )
+
+    state_count_pcoll = (
+        exp_model_pcoll
+        | 'Count states' >> beam.Map(self.get_number_of_states)
+    )
+
+    state_count_sum_pcoll = (
+        state_count_pcoll
+        | 'Sum values' >> beam.CombineGlobally(sum)
+    )
+
+    return state_count_sum_pcoll | beam.Map(job_run_result.JobRunResult.as_stdout)
+```
+
+The method maps every element in a `PCollection` into a `JobRunResult` with the stringified-value as its `stdout`.
+
+With this, our job is fully written! Move on to the next section for instructions on _testing_ your new job.
+
+## Testing Apache Beam Jobs
+
+We'll assume that we've written a fully operational `CountExplorationStatesJob`, and go over the steps in ensuring the job works as intended.
+
+### 1. Inherit from `JobTestBase` and override the class constant `JOB_CLASS`
+
+The current convention is to name your test cases `<JobName>Tests`, but you can create better names if you want to break tests up into groups of specialized focus. For our example, we'll keep things simple.
+
+```python
+class CountExplorationStatesJobTests(test_jobs.JobTestBase):
+
+    JOB_CLASS = CountExplorationStatesJob
+```
+
+### 2. Run assertions using a `assert_job_output_is_*` method
```

Comment:
Explain best practices for figuring out which tests to write.

Explain what sorts of things can and cannot be tested. How will we verify the parts that cannot be tested through this approach?

Reply:
Done.

Reply:
Hm. How about the second part? 

> Explain what sorts of things can and cannot be tested. How will we verify the parts that cannot be tested through this approach?

What do these tests guarantee -- will they guarantee a successful and correct prod run? If not, what else is needed?

Reply:
How's this?